### PR TITLE
Debug info improvements for Mandrel

### PR DESCRIPTION
**Pull request by galderz**
Replaces #10759. 

This PR targets Quarkus 1.8 to avoid last minute changes with Quarkus 1.7 release.

### Third party sources

Add third party source jars support. 

Third party jars not downloaded by default, so it requires user to call `mvn dependencies:sources` or equivalent. Checked with @aloubyansky and we might want to automate this extra step somehow. I'll create a separate issue to track this.

If source jars are found, they're put next to the jars in the lib directory. After generating native executable, the source jars are removed, so it has no visible effect. Having the source jars next to the jars means they're automatically detected.

### Application sources added to the sources cache

Application sources are added individually using `-H:DebugInfoSourceSearchPath` option. They're only added if the directory exists.

### Move the sources cache folder to be under `target/sources-cache`

Sources cache has been moved to `target/sources-cache` using the `-H:DebugInfoSourceCacheRoot` option. This makes sources cache more closely available. `gdb` should discover these automatically, but it does not seem to work, so better use `set directories` pointing to them. I'm working with @zakkak to see what's up there.
 
### GraalVM version and distribution resolution.

These changes only work on Mandrel or GraalVM 20.2, so reworked the whole GraalVM version and distribution resolution logic to make this easier to handle. Previous ad-hoc GraalVM version resolution has been been replaced by a type-safe regex based solution. We can now compare versions to enable things only for versions newer than X.

Added unit tests to verify that GraalVM versions and distributions are resolved correctly with a diverse set of `native-image --version` outputs. If there are any issues we can just add other outputs and see what happens.

### Others

Debug symbols flag switched from -H:GenerateDebugInfo=1 to `-g`.

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 1]
```diff
@@ -572,4 +646,104 @@ private static void objcopy(String... args) {
             }
         }
     }
+
+    protected static final class GraalVM {
+        static final class Version implements Comparable<Version> {
+            private static final Pattern PATTERN = Pattern.compile(
+                    "GraalVM Version ((1|[1-9][0-9]).([0-3]).[0-9]|\\p{XDigit}*)[^(\n$]*(\\(Mandrel Distribution\\))?\\s*");
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:653]: The pattern is not good enough IMHO. I will push a commit to fix it.

**Comment by gsmet** [comment]
I rebased and added a commit to fix the regexp. There were several issues: we shouldn't consider minor version can only be `<= 3`, that won't scale. `\.` should be used to catch dots not `.` as it's the wildcard character.

**Review by gsmet — APPROVED** [review]

**Comment by galderz** [comment]
@gsmet Thanks for changes.

FYI, according to GraalVM, there will only be 4 minor releases each year, see [here](https://www.graalvm.org/docs/release-notes/version-roadmap/). Hence limiting to 0-3 seemed fine to me.

**Comment by zakkak** [comment]
@galderz according to https://www.graalvm.org/docs/release-notes/version-roadmap/ this has now been made 1-4 so you get:
* 19.3
* 19.3.1
* 19.3.2
* 19.3.3
* 19.3.4

**Comment by gsmet** [comment]
Well, limiting it to 0-3 really doesn't bring anything except the risk to
be broken at some point. That's why it's better to get rid of it anyway.

**Comment by galderz** [comment]
@zakkak I was talking about minor, not micro

**Comment by galderz** [comment]
As in: `major.minor.micro`

**Comment by zakkak** [comment]
Ah right, \me should think better before speaking :)

**Pull request merged**